framework/src/media/audio/audio_manager.c

```diff
@@ -1110,3 +1227,260 @@ audio_manager_result_t set_output_audio_volume(uint8_t volume)
 {
 	return set_audio_volume(OUTPUT, volume);
 }
+
+static audio_manager_result_t get_supported_process_type(int fd, audio_config_t *config, audio_io_direction_t direct)
+{
+	struct audio_caps_desc_s caps_desc;
+	pthread_mutex_t *card_mutex;
+
+	if (direct == OUTPUT) {		// TODO add output logic in future.
+		meddbg("Do not support output for now...\n");
+		return AUDIO_MANAGER_SUCCESS;
+	} else {
+		card_mutex = &g_audio_in_cards[g_actual_audio_in_card_id].card_mutex;
+	}
+
+	caps_desc.caps.ac_len = sizeof(struct audio_caps_s);
+	caps_desc.caps.ac_type = AUDIO_TYPE_PROCESSING;
+	caps_desc.caps.ac_subtype = AUDIO_PU_UNDEF;
+	pthread_mutex_lock(card_mutex);
+	if (ioctl(fd, AUDIOIOC_GETCAPS, (unsigned long)&caps_desc.caps) < 0) {
+		meddbg("An ioctl error occurs\n");
+		pthread_mutex_unlock(card_mutex);
+		return AUDIO_MANAGER_DEVICE_FAIL;
+	}
+
+	config->device_process_type = caps_desc.caps.ac_controls.b[0];
+	pthread_mutex_unlock(card_mutex);
+	medvdbg("process type = 0x%x\n", config->device_process_type);
+
+	return AUDIO_MANAGER_SUCCESS;
+}
+
+/* TODO below functions should be changed,
+  * currently we using static array for cards, but below should be modified when we change to handler structure */
+audio_manager_result_t find_stream_in_device_with_process_type(device_process_type_t type, int *card_id, int *device_id, device_process_subtype_t *subtype)
+{
+	int i;
+	int fd;
+	audio_manager_result_t ret;
+	char path[AUDIO_PCM_FULL_PATH_LENGTH];
+	struct audio_caps_desc_s caps_desc;
+
+	ret = init_audio_stream(INPUT);
+	if (ret != AUDIO_MANAGER_SUCCESS) {
+		meddbg("init stream failed\n");
+		return ret;
+	}
+
+	caps_desc.caps.ac_len = sizeof(struct audio_caps_s);
+	caps_desc.caps.ac_type = AUDIO_TYPE_PROCESSING;
+	caps_desc.caps.ac_subtype = type;
+
+	for (i = 0; i < CONFIG_AUDIO_MAX_DEVICE_NUM; i++) {
+		if (((g_audio_in_cards[g_actual_audio_in_card_id].config[i].device_process_type) & type) != 0) {
+			meddbg("found!!device_id : %d card_id : %d device : %x given : %x\n", i, g_actual_audio_in_card_id, g_audio_in_cards[g_actual_audio_in_card_id].config[i].device_process_type, type);
+			get_card_path(path, g_actual_audio_in_card_id, i, INPUT);
+			pthread_mutex_lock(&(g_audio_in_cards[g_actual_audio_in_card_id].card_mutex));
+			fd = open(path, O_RDONLY);
+			if (fd < 0) {
+				meddbg("open failed, path : %s process type : 0x%x\n", path, type);
+				return AUDIO_MANAGER_NO_AVAIL_CARD;
```

Comment:
mutex_unlock is required before returning.

Reply:
This function will be modified totally.